### PR TITLE
feat(sourcing): support bundle products in default sourcing service

### PR DESCRIPTION
**Pull request by KarolNowakk**

### CODE REVIEW

File: sourcing/domain/service.go [hunk 1]
```diff
@@ -167,27 +178,76 @@ func (d *DefaultSourcingService) AllocateItems(ctx context.Context, decoratedCar
...
+
+	itemAllocation := ItemAllocation{
+		AllocatedQtys: allocatedQtys,
+		Error:         err,
+	}
+
+	return itemAllocation, updatedSourcestock, nil
+}
+
+func (d *DefaultSourcingService) allocateBundleWithActiveChoices(ctx context.Context, itemQty int, productSourcestock map[string]map[Source]int, bundleProduct domain.BundleProductWithActiveChoices, deliveryInfo cartDomain.DeliveryInfo) (ItemAllocation, map[string]map[Source]int) {
```
Comment (IvanMaidurov) [inline sourcing/domain/service.go:228]: very minor: you can place arguments on separate lines to make it easier to read

File: sourcing/domain/service.go [hunk 2]
```diff
@@ -210,78 +270,76 @@ func getItemIdsWithProduct(dc *decorator.DecoratedCart, product domain.BasicProd
...
-func (d *DefaultSourcingService) allocateItem(ctx context.Context, productSourcestock map[string]map[Source]int, decoratedItem decorator.DecoratedCartItem, deliveryInfo cartDomain.DeliveryInfo) (ItemAllocation, map[string]map[Source]int) {
-	var resultItemAllocation = ItemAllocation{
-		AllocatedQtys: make(AllocatedQtys),
-	}
-	// copy given known stock
-	remainingSourcestock := productSourcestock
-
-	productID := decoratedItem.Product.GetIdentifier()
+func (d *DefaultSourcingService) allocateItem(ctx context.Context, productSourcestock map[string]map[Source]int, product domain.BasicProduct, qtyToAllocate int, deliveryInfo cartDomain.DeliveryInfo) (AllocatedQtys, map[string]map[Source]int, error) {
+	productID := product.GetIdentifier()
```
Comment (IvanMaidurov) [inline sourcing/domain/service.go:278]: I guess you don't need productID here anymore. You can validate product.GetIdentifier() directly and then not pass product id to d.allocateFromSources()

File: sourcing/domain/service.go [hunk 3]
```diff
@@ -167,27 +207,87 @@ func (d *DefaultSourcingService) AllocateItems(ctx context.Context, decoratedCar
...
+	return resultItemAllocations, nil
+}
+
+func (d *DefaultSourcingService) allocateItemForProduct(
+	ctx context.Context,
+	productSourcestock map[string]map[Source]int,
+	decoratedItem *decorator.DecoratedCartItem,
+	deliveryInfo cartDomain.DeliveryInfo,
+) (ItemAllocation, map[string]map[Source]int, error) {
+	if decoratedItem.Product.Type() == domain.TypeBundle || decoratedItem.Product.Type() == domain.TypeConfigurableWithActiveVariant {
```
Comment (carstendietrich) [inline sourcing/domain/service.go:243]: why cant we source configurable with active variant?

File: sourcing/domain/service.go [hunk 4]
```diff
@@ -210,78 +310,100 @@ func getItemIdsWithProduct(dc *decorator.DecoratedCart, product domain.BasicProd
...
+func (d *DefaultSourcingService) allocateItem(
+	ctx context.Context,
+	productSourcestock map[string]map[Source]int,
+	product domain.BasicProduct,
+	qtyToAllocate int,
+	deliveryInfo cartDomain.DeliveryInfo,
+) (AllocatedQtys, map[string]map[Source]int, error) {
+	sources, err := d.availableSourcesProvider.GetPossibleSources(ctx, product, &deliveryInfo)
+	if err != nil {
+		return nil, productSourcestock, fmt.Errorf("error getting possible sources: %w", err)
```
Comment (carstendietrich) [inline sourcing/domain/service.go:326]: maybe add the product identifier which we try to source for better errors

**Review by carstendietrich — CHANGES_REQUESTED** [review]
Please mention changes in Changelog.

File: sourcing/domain/service.go [hunk 5]
```diff
@@ -109,16 +127,61 @@ func (d *DefaultSourcingService) Inject(
...
+			}
+
+			availableSourceForBundle[ProductID(choice.Product.GetIdentifier())] = qtys
+		}
+
+		return availableSourceForBundle, nil
+	}
+
+	// check here so we don't need to check further
+	if product.GetIdentifier() == "" {
```
Comment (carstendietrich) [inline sourcing/domain/service.go:166]: maybe move that at the beginning of the func like the checks for type etc

File: sourcing/domain/service.go [hunk 6]
```diff
@@ -144,144 +207,243 @@ func (d *DefaultSourcingService) GetAvailableSources(ctx context.Context, produc
...
+		}
+
+		resultItemAllocation.AllocatedQtys[ProductID(activeChoice.Product.GetIdentifier())] = allocatedQtys
 	}
-	sources, err := d.availableSourcesProvider.GetPossibleSources(ctx, decoratedItem.Product, &deliveryInfo)
+
+	return resultItemAllocation
+}
+
+func (d *DefaultSourcingService) allocateItem(
```
Comment (carstendietrich) [inline sourcing/domain/service.go:365]: maybe rename to `allocateProduct`

File: sourcing/domain/service.go [hunk 7]
```diff
@@ -144,144 +207,243 @@ func (d *DefaultSourcingService) GetAvailableSources(ctx context.Context, produc
...
 
-	return resultItemAllocations, overallError
+	return resultItemAllocations, nil
 }
 
-func (d *DefaultSourcingService) checkConfiguration() error {
-	if d.availableSourcesProvider == nil {
-		d.logger.Error("no Source Provider bound")
-		return errors.New("no Source Provider bound")
+func (d *DefaultSourcingService) allocateItemForProduct(
```
Comment (carstendietrich) [inline sourcing/domain/service.go:297]: would name this just allocateItem

**Review by carstendietrich — CHANGES_REQUESTED** [review]

**Review by carstendietrich — CHANGES_REQUESTED** [review]

**Review by carstendietrich — APPROVED** [review]

**Pull request merged**